### PR TITLE
[AIX] Handle toc-data offset overflowing 16-bits

### PR DESCRIPTION
**Pull request by syzaara**
When the toc-data offset overflows the 16-bits, we can truncate the value to the 16-bit value as the linker will handle overflow through fixup code.

### CODE REVIEW

File: llvm/test/CodeGen/PowerPC/aix-overflow-toc-data.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+# RUN: %python %s > %t.ll
+# RUN: llc -mtriple powerpc-ibm-aix-xcoff -code-model=small -mcpu=pwr4 -mattr=-altivec -O0 < %t.ll | \
+# RUN:   FileCheck --check-prefix=ASM32 %s
+
+# RUN: llc -mtriple powerpc64-ibm-aix-xcoff -code-model=small -mcpu=pwr4 -mattr=-altivec -O0 < %t.ll | \
+# RUN:   FileCheck --check-prefix=ASM64 %s
+
+# RUN: llc -mtriple powerpc-ibm-aix-xcoff -code-model=small -mcpu=pwr4 -mattr=-altivec -O0 \
+# RUN:     -filetype=obj -o %t.o < %t.ll
+# RUN: llvm-objdump --no-print-imm-hex -D -r --symbol-description %t.o | FileCheck --check-prefix=DIS32 %s
```
Comment (amy-kwan) [inline llvm/test/CodeGen/PowerPC/aix-overflow-toc-data.py:12]: Does this test case need a change similar to https://github.com/llvm/llvm-project/pull/80162/files#diff-1d8042ef4a587126a03f4ef2eca2e9a7366c90db225e78665f01ae777a9436ab?

**Review by amy-kwan — APPROVED** [review]
Thanks for addressing my comment. I do not have any additional comments so unless there are some from others, LGTM.

File: llvm/test/CodeGen/PowerPC/aix-overflow-toc-data.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# UNSUPPORTED: expensive_checks, debug
+
+# RUN: %python %s > %t.ll
+# RUN: llc -mtriple powerpc-ibm-aix-xcoff -code-model=small -mcpu=pwr4 -mattr=-altivec -O0 < %t.ll | \
```
Comment (mandlebug) [inline llvm/test/CodeGen/PowerPC/aix-overflow-toc-data.py:4]: I believe for new tests we target pwr7 which is the default arch for AIX.

File: llvm/lib/MC/XCOFFObjectWriter.cpp [hunk 3]
```diff
@@ -736,11 +736,26 @@ void XCOFFObjectWriter::recordRelocation(MCAssembler &Asm,
...
+      // various TOC PseudoOps.
+      // For toc-data symbols, we were not able to calculate the offset from
+      // the TOC in PPCAsmPrinter.cpp since the TOC has not been finalized at
+      // that point, so we are adjusting it here though
+      // llvm::SignExtend64<16>(TOCEntryOffset);
+      // TODO: Since the time that the handling for offsets over 16-bits was
+      // added in PPCAsmPrinter.cpp using getTOCEntryLoadingExprForXCOFF, the
+      // system assembler and linker have been updated to be able to handle the
+      // overflowing offsets, so we no longer need to keep
+      // getTOCEntryLoadingExprForXCOFF.
```
[the remark this answers is not in the document]
Comment (syzaara) [reply llvm/lib/MC/XCOFFObjectWriter.cpp:756]: Further investigation is needed to verify the various supported OS levels to check if the fix is available. We can do that separately to this patch and open up follow up patch to remove the truncation in PPCAsmPrinter.cpp once verified.

**Pull request merged**